### ISSUE TITLE
The saved project contains dead data

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1184]
- **Version:** Git Master Branch
- **OS:** Windows

Repro project: [Bugs.zip](https://github.com/OurMachinery/themachinery-public/files/8283953/Bugs.zip)

This project is a stripped down version of our old big machinery project, saved with the-machinery built with latest Git Master Branch pulled 17 March 2022.

Open the project in the-machinery and note: The "world.entity" only have a prototype which it has no overrides for, and one child asset entity where the transform is overridden (this is the default state when adding a child entity from asset, not actively chosen)

Now check what the world.tm_entity file on disk contains:
- One "components__removed" is saved, but there is no component removed in the editor
- Two "components__instantiated" are saved, but there are no components added
- One "children__instantiated", but there is no child entity in the world. This non-existant child entity has a `tm_physics_shape_heightfield` with `tm_physx_heightfield_cooked_data` which is saved on disk, 64MB big. Stored but not used.
- A long list of child_sort_values for children which doesn't exist.
- This might be nothing: The `validity_hash` of `__asset_thumbnail` is `ffffffffffffffff` which sounds suspect?

**Comment by niklas-ourmachinery** [comment]
Yes, we don't clear out the `removed` and `instantiated` data when the original object they refer to is removed from the prototype. We should probably do that on save.

I'm not 100 % sure about the `child_sort_values`... looks like they should be removed when the child is deleted, but there may be some cases where it doesn't happen. @leonardo-ourmachinery ?

**Comment by leonardo-ourmachinery** [comment]
The problem with the `child_sort_values` is that they are computed in "retained mode", meaning that the entity tree recomputes the list from scratch if either you reorder entities or if you add new entities to it. 
This works fine for simple cases where you just press "delete" (the child sort value is removed immediately) but it doesn't work for more complex case like you override and then un-override a child or when you undo/redo some operations that add/remove children.

I made it so that when we normalize the children list we also clear up the `child_sort_values` list of the parent.
Unfortunately in this particular case the dead-data will only go away once a children is added/removed to the world entity, as otherwise we won't wipe out the sort value list.

**Comment by simon-ourmachinery** [comment]
@karl-zylinski-ourmachinery and @tobias-ourmachinery  I have noticed that the graph can leak quite some dead data: Especially when using dynamic connectors such as on subgraph nodes or on input/output nodes. My solution for the variable nodes is to loop over all data and connections in the graph and remove them if the variable is removed.
